File: ios/Beanshelf/Data/SocialClient.swift
import UIKit

/// Client for the Beanshelf social server (server/main.py in this repo).
/// URLSession + JSONSerialization — no dependencies. Sign-in state (server
/// URL, token, username) lives in UserDefaults. Same API as the Android app;
/// both platforms talk to the same server.
enum SocialClient {

    static let defaultServer = "https://beans.beanshelf.ca"

    struct Account: Codable {
        var serverUrl: String
        var token: String
        var username: String
        var display: String
    }

    struct FeedPost: Identifiable {
        let id: String
        let username: String
        let display: String
        let name: String
        let roaster: String
        let origin: String
        let variety: String
        let process: String
        let notes: String
        let rating: Double
        let photoUrl: URL? // absolute
        let createdAt: Int64
        let cheers: Int
        let iCheered: Bool
        let commentCount: Int
        let lastCommentUser: String?
        let lastCommentText: String?
    }

    struct UserHit: Identifiable {
        let username: String
        let display: String
        var following: Bool
        var id: String { username }
    }

    struct Profile {
        let username: String
        let display: String
        let followers: Int
        let following: Int
        let beans: Int
        let iFollow: Bool
        let isMe: Bool
        let profileUrl: String
    }

    struct Comment: Identifiable {
        let id: String
        let username: String
        let display: String
        let text: String
        let createdAt: Int64
    }

    struct SocialError: LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    private static let accountKey = "social.account"
    private static let lastServerKey = "social.lastServerUrl"

    static var account: Account? {
        guard let data = UserDefaults.standard.data(forKey: accountKey) else { return nil }
        return try? JSONDecoder().decode(Account.self, from: data)
    }

    static var lastServerUrl: String {
        UserDefaults.standard.string(forKey: lastServerKey) ?? ""
    }

    static func signOut() {
        UserDefaults.standard.removeObject(forKey: accountKey)
    }

    private static func save(_ a: Account) {
        if let data = try? JSONEncoder().encode(a) {
            UserDefaults.standard.set(data, forKey: accountKey)
        }
        UserDefaults.standard.set(a.serverUrl, forKey: lastServerKey)
    }

    // ── HTTP plumbing ─────────────────────────────────────────────────────
    private static func request(
        base: String,
        path: String,
        method: String,
        token: String?,
        body: [String: Any]?
    ) async throws -> Data {
        let trimmed = base.hasSuffix("/") ? String(base.dropLast()) : base
        guard let url = URL(string: trimmed + path) else {
            throw SocialError(message: "That server address doesn't look right")
        }
        var req = URLRequest(url: url, timeoutInterval: 15)
        req.httpMethod = method
        if let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try? JSONSerialization.data(withJSONObject: body)
        }
        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await URLSession.shared.data(for: req)
        } catch {
            throw SocialError(message: "Couldn't reach the server")
        }
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200...299).contains(code) else {
            let detail = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])
                .flatMap { $0["detail"] as? String }
            throw SocialError(message: (detail?.isEmpty == false ? detail! : "Server error (\(code))"))
        }
        return data
    }

    // ── auth ──────────────────────────────────────────────────────────────
    static func register(serverUrl: String, username: String, display: String, password: String) async throws -> Account {
        try await authCall(serverUrl: serverUrl, path: "/auth/register", username: username, display: display, password: password)
    }

    static func login(serverUrl: String, username: String, password: String) async throws -> Account {
        try await authCall(serverUrl: serverUrl, path: "/auth/login", username: username, display: "", password: password)
    }

    private static func authCall(
        serverUrl: String, path: String,
        username: String, display: String, password: String
    ) async throws -> Account {
        let data = try await request(base: serverUrl, path: path, method: "POST", token: nil, body: [
            "username": username.trimmingCharacters(in: .whitespaces).lowercased(),
            "password": password,
            "display": display.trimmingCharacters(in: .whitespaces),
        ])
        guard let res = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = res["token"] as? String,
              let user = res["username"] as? String
        else { throw SocialError(message: "Unexpected reply from the server") }
        let trimmed = serverUrl.hasSuffix("/") ? String(serverUrl.dropLast()) : serverUrl
        let account = Account(
            serverUrl: trimmed,
            token: token,
            username: user,
            display: (res["display"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? user
        )
        save(account)
        return account
    }

    // ── social ────────────────────────────────────────────────────────────
    static func search(_ a: Account, query: String) async throws -> [UserHit] {
        let q = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query
        let data = try await request(base: a.serverUrl, path: "/users/search?q=\(q)", method: "GET", token: a.token, body: nil)
        let arr = (try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]) ?? []
        return arr.compactMap { o in
            guard let username = o["username"] as? String else { return nil }
            return UserHit(
                username: username,
                display: o["display"] as? String ?? username,
                following: o["following"] as? Bool ?? false
            )
        }
    }

    static func setFollowing(_ a: Account, username: String, follow: Bool) async throws {
        _ = try await request(
            base: a.serverUrl,
            path: "/follow/\(username)",
            method: follow ? "POST" : "DELETE",
            token: a.token,
            body: follow ? [:] : nil
        )
    }

    static func feed(_ a: Account) async throws -> [FeedPost] {
        parsePosts(a, try await request(base: a.serverUrl, path: "/feed", method: "GET", token: a.token, body: nil))
    }

    static func discover(_ a: Account) async throws -> [FeedPost] {
        parsePosts(a, try await request(base: a.serverUrl, path: "/discover", method: "GET", token: a.token, body: nil))
    }

    static func userLeaderboard(_ a: Account, username: String) async throws -> [FeedPost] {
        parsePosts(a, try await request(base: a.serverUrl, path: "/users/\(username)/leaderboard", method: "GET", token: a.token, body: nil))
    }

    static func profile(_ a: Account, username: String) async throws -> Profile {
        let data = try await request(base: a.serverUrl, path: "/users/\(username)/profile", method: "GET", token: a.token, body: nil)
        guard let o = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let user = o["username"] as? String
        else { throw SocialError(message: "Unexpected reply from the server") }
        return Profile(
            username: user,
            display: o["display"] as? String ?? user,
            followers: o["followers"] as? Int ?? 0,
            following: o["following"] as? Int ?? 0,
            beans: o["beans"] as? Int ?? 0,
            iFollow: o["iFollow"] as? Bool ?? false,
            isMe: o["isMe"] as? Bool ?? false,
            profileUrl: o["profileUrl"] as? String ?? ""
        )
    }

    private static func parsePeople(_ data: Data) -> [UserHit] {
        let arr = (try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]) ?? []
        return arr.compactMap { o in
            guard let username = o["username"] as? String else { return nil }
            return UserHit(
                username: username,
                display: o["display"] as? String ?? username,
                following: o["following"] as? Bool ?? false
            )
        }
    }

    static func followers(_ a: Account, username: String) async throws -> [UserHit] {
        parsePeople(try await request(base: a.serverUrl, path: "/users/\(username)/followers", method: "GET", token: a.token, body: nil))
    }

    static func followingList(_ a: Account, username: String) async throws -> [UserHit] {
        parsePeople(try await request(base: a.serverUrl, path: "/users/\(username)/following", method: "GET", token: a.token, body: nil))
    }

    /// Returns the new cheer count.
    static func setCheer(_ a: Account, postId: String, on: Bool) async throws -> Int {
        let data = try await request(
            base: a.serverUrl,
            path: "/beans/\(postId)/cheers",
            method: on ? "POST" : "DELETE",
            token: a.token,
            body: on ? [:] : nil
        )
        let o = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        return o?["cheers"] as? Int ?? 0
    }

    static func comments(_ a: Account, postId: String) async throws -> [Comment] {
        let data = try await request(base: a.serverUrl, path: "/beans/\(postId)/comments", method: "GET", token: a.token, body: nil)
        let arr = (try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]) ?? []
        return arr.compactMap { o in
            guard let id = o["id"] as? String else { return nil }
            return Comment(
                id: id,
                username: o["username"] as? String ?? "",
                display: o["display"] as? String ?? "",
                text: o["text"] as? String ?? "",
                createdAt: (o["createdAt"] as? NSNumber)?.int64Value ?? 0
            )
        }
    }

    static func addComment(_ a: Account, postId: String, text: String) async throws {
        _ = try await request(base: a.serverUrl, path: "/beans/\(postId)/comments", method: "POST", token: a.token, body: ["text": text])
    }

    private static func parsePosts(_ a: Account, _ data: Data) -> [FeedPost] {
        let arr = (try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]) ?? []
        return arr.compactMap { o in
            guard let id = o["id"] as? String, let username = o["username"] as? String else { return nil }
            return FeedPost(
                id: id,
                username: username,
                display: o["display"] as? String ?? username,
                name: o["name"] as? String ?? "",
                roaster: o["roaster"] as? String ?? "",
                origin: o["origin"] as? String ?? "",
                variety: o["variety"] as? String ?? "",
                process: o["process"] as? String ?? "",
                notes: o["notes"] as? String ?? "",
                rating: o["rating"] as? Double ?? 0,
                photoUrl: (o["photoUrl"] as? String).flatMap { URL(string: a.serverUrl + $0) },
                createdAt: (o["createdAt"] as? NSNumber)?.int64Value ?? 0,
                cheers: o["cheers"] as? Int ?? 0,
                iCheered: o["iCheered"] as? Bool ?? false,
                commentCount: o["commentCount"] as? Int ?? 0,
                lastCommentUser: o["lastCommentUser"] as? String,
                lastCommentText: o["lastCommentText"] as? String
            )
        }
    }

    /// Posts a bean check-in with its photo downscaled to ~800px.
    static func postBean(_ a: Account, bean: Bean) async throws {
        var body: [String: Any] = [
            "name": bean.name,
            "roaster": bean.roaster,
            "origin": bean.origin,
            "variety": bean.variety,
            "process": bean.process,
            "notes": bean.notes,
            "rating": bean.rating,
        ]
        if let file = bean.photoFile,
           let url = PhotoStore.url(for: file),
           let image = UIImage(contentsOfFile: url.path) {
            let pixelW = image.size.width * image.scale
            let pixelH = image.size.height * image.scale
            let scale = min(1, 800 / max(pixelW, pixelH))
            let target = CGSize(width: pixelW * scale, height: pixelH * scale)
            let format = UIGraphicsImageRendererFormat()
            format.scale = 1
            // keep PNG for cutouts so transparency survives into the feed
            let isCutout = BagCropper.isCutout(file)
            format.opaque = !isCutout
            let scaled = UIGraphicsImageRenderer(size: target, format: format).image { _ in
                image.draw(in: CGRect(origin: .zero, size: target))
            }
            let data = isCutout ? scaled.pngData() : scaled.jpegData(compressionQuality: 0.82)
            if let data { body["photo_b64"] = data.base64EncodedString() }
        }
        _ = try await request(base: a.serverUrl, path: "/beans", method: "POST", token: a.token, body: body)
    }

    /// Fetch + cache a feed photo (keyed by post id) for display.
    static func fetchPhoto(post: FeedPost) async -> UIImage? {
        guard let url = post.photoUrl else { return nil }
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent("feed", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let ext = url.path.hasSuffix(".png") ? "png" : "jpg"
        let cached = dir.appendingPathComponent("\(post.id).\(ext)")
        if let data = try? Data(contentsOf: cached), let image = UIImage(data: data) {
            return image
        }
        guard let (data, _) = try? await URLSession.shared.data(from: url),
              let image = UIImage(data: data) else { return nil }
        try? data.write(to: cached, options: .atomic)
        return image
    }
}

func relativeTime(_ ts: Int64) -> String {
    let mins = (nowMillis() - ts) / 60000
    switch mins {
    case ..<1: return "now"
    case ..<60: return "\(mins)m"
    case ..<(60 * 24): return "\(mins / 60)h"
    default: return "\(mins / (60 * 24))d"
    }
}

File: app/src/main/java/dev/adamsjack/beanshelf/data/SocialClient.kt
package dev.adamsjack.beanshelf.data

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import dev.adamsjack.beanshelf.model.Bean
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.net.HttpURLConnection
import java.net.URL

/**
 * Client for the Beanshelf social server (server/main.py). Plain
 * HttpURLConnection + org.json — no networking deps. All calls suspend on IO.
 * Sign-in state (server URL, token, username) lives in SharedPreferences.
 */
object SocialClient {

    data class Account(val serverUrl: String, val token: String, val username: String, val display: String)
    data class FeedPost(
        val id: String,
        val username: String,
        val display: String,
        val name: String,
        val roaster: String,
        val origin: String,
        val variety: String,
        val process: String,
        val notes: String,
        val rating: Float,
        val photoUrl: String?, // absolute
        val createdAt: Long,
        val cheers: Int,
        val iCheered: Boolean,
        val commentCount: Int,
        val lastCommentUser: String?,
        val lastCommentText: String?,
    )
    data class UserHit(val username: String, val display: String, val following: Boolean)
    data class Profile(
        val username: String,
        val display: String,
        val followers: Int,
        val following: Int,
        val beans: Int,
        val iFollow: Boolean,
        val isMe: Boolean,
        val profileUrl: String,
    )
    data class Comment(val id: String, val username: String, val display: String, val text: String, val createdAt: Long)

    class SocialException(message: String) : Exception(message)

    private const val PREFS = "social"

    fun account(context: Context): Account? {
        val p = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val url = p.getString("serverUrl", null) ?: return null
        val token = p.getString("token", null) ?: return null
        val username = p.getString("username", null) ?: return null
        return Account(url, token, username, p.getString("display", username) ?: username)
    }

    fun signOut(context: Context) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().clear().apply()
    }

    fun lastServerUrl(context: Context): String =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString("lastServerUrl", "") ?: ""

    private fun saveAccount(context: Context, a: Account) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putString("serverUrl", a.serverUrl)
            .putString("lastServerUrl", a.serverUrl)
            .putString("token", a.token)
            .putString("username", a.username)
            .putString("display", a.display)
            .apply()
    }

    // ── HTTP plumbing ─────────────────────────────────────────────────────
    private suspend fun request(
        base: String,
        path: String,
        method: String,
        token: String?,
        body: JSONObject?,
    ): String = withContext(Dispatchers.IO) {
        val conn = URL(base.trimEnd('/') + path).openConnection() as HttpURLConnection
        try {
            conn.requestMethod = method
            conn.connectTimeout = 6000
            conn.readTimeout = 15000
            token?.let { conn.setRequestProperty("Authorization", "Bearer $it") }
            if (body != null) {
                conn.setRequestProperty("Content-Type", "application/json")
                conn.doOutput = true
                conn.outputStream.use { it.write(body.toString().toByteArray()) }
            }
            val code = conn.responseCode
            val text = (if (code in 200..299) conn.inputStream else conn.errorStream)
                ?.use { it.readBytes().toString(Charsets.UTF_8) } ?: ""
            if (code !in 200..299) {
                val detail = runCatching { JSONObject(text).optString("detail") }.getOrNull()
                throw SocialException(detail?.ifBlank { null } ?: "Server error ($code)")
            }
            text
        } finally {
            conn.disconnect()
        }
    }

    // ── auth ──────────────────────────────────────────────────────────────
    suspend fun register(context: Context, serverUrl: String, username: String, display: String, password: String): Account =
        authCall(context, serverUrl, "/auth/register", username, display, password)

    suspend fun login(context: Context, serverUrl: String, username: String, password: String): Account =
        authCall(context, serverUrl, "/auth/login", username, "", password)

    private suspend fun authCall(
        context: Context, serverUrl: String, path: String,
        username: String, display: String, password: String,
    ): Account {
        val res = JSONObject(
            request(serverUrl, path, "POST", null, JSONObject().apply {
                put("username", username.trim().lowercase())
                put("password", password)
                put("display", display.trim())
            })
        )
        val account = Account(
            serverUrl = serverUrl.trimEnd('/'),
            token = res.getString("token"),
            username = res.getString("username"),
            display = res.optString("display", username),
        )
        saveAccount(context, account)
        return account
    }

    // ── social ────────────────────────────────────────────────────────────
    suspend fun search(a: Account, query: String): List<UserHit> {
        val arr = JSONArray(request(a.serverUrl, "/users/search?q=" + java.net.URLEncoder.encode(query, "UTF-8"), "GET", a.token, null))
        return (0 until arr.length()).map { i ->
            val o = arr.getJSONObject(i)
            UserHit(o.getString("username"), o.optString("display"), o.optBoolean("following"))
        }
    }

    suspend fun setFollowing(a: Account, username: String, follow: Boolean) {
        request(a.serverUrl, "/follow/$username", if (follow) "POST" else "DELETE", a.token, JSONObject().takeIf { follow })
    }

    suspend fun feed(a: Account): List<FeedPost> = parsePosts(a, request(a.serverUrl, "/feed", "GET", a.token, null))

    suspend fun discover(a: Account): List<FeedPost> = parsePosts(a, request(a.serverUrl, "/discover", "GET", a.token, null))

    suspend fun userLeaderboard(a: Account, username: String): List<FeedPost> =
        parsePosts(a, request(a.serverUrl, "/users/$username/leaderboard", "GET", a.token, null))

    suspend fun profile(a: Account, username: String): Profile {
        val o = JSONObject(request(a.serverUrl, "/users/$username/profile", "GET", a.token, null))
        return Profile(
            username = o.getString("username"),
            display = o.optString("display"),
            followers = o.optInt("followers"),
            following = o.optInt("following"),
            beans = o.optInt("beans"),
            iFollow = o.optBoolean("iFollow"),
            isMe = o.optBoolean("isMe"),
            profileUrl = o.optString("profileUrl"),
        )
    }

    private fun parsePeople(text: String): List<UserHit> {
        val arr = JSONArray(text)
        return (0 until arr.length()).map { i ->
            val o = arr.getJSONObject(i)
            UserHit(o.getString("username"), o.optString("display"), o.optBoolean("following"))
        }
    }

    suspend fun followers(a: Account, username: String): List<UserHit> =
        parsePeople(request(a.serverUrl, "/users/$username/followers", "GET", a.token, null))

    suspend fun followingList(a: Account, username: String): List<UserHit> =
        parsePeople(request(a.serverUrl, "/users/$username/following", "GET", a.token, null))

    /** Returns the new cheer count. */
    suspend fun setCheer(a: Account, postId: String, on: Boolean): Int {
        val o = JSONObject(request(a.serverUrl, "/beans/$postId/cheers", if (on) "POST" else "DELETE", a.token, JSONObject().takeIf { on }))
        return o.optInt("cheers")
    }

    suspend fun comments(a: Account, postId: String): List<Comment> {
        val arr = JSONArray(request(a.serverUrl, "/beans/$postId/comments", "GET", a.token, null))
        return (0 until arr.length()).map { i ->
            val o = arr.getJSONObject(i)
            Comment(o.getString("id"), o.optString("username"), o.optString("display"), o.optString("text"), o.optLong("createdAt"))
        }
    }

    suspend fun addComment(a: Account, postId: String, text: String) {
        request(a.serverUrl, "/beans/$postId/comments", "POST", a.token, JSONObject().apply { put("text", text) })
    }

    private fun parsePosts(a: Account, text: String): List<FeedPost> {
        val arr = JSONArray(text)
        return (0 until arr.length()).map { i ->
            val o = arr.getJSONObject(i)
            FeedPost(
                id = o.getString("id"),
                username = o.getString("username"),
                display = o.optString("display"),
                name = o.optString("name"),
                roaster = o.optString("roaster"),
                origin = o.optString("origin"),
                variety = o.optString("variety"),
                process = o.optString("process"),
                notes = o.optString("notes"),
                rating = o.optDouble("rating", 0.0).toFloat(),
                photoUrl = if (o.isNull("photoUrl")) null else a.serverUrl + o.getString("photoUrl"),
                createdAt = o.optLong("createdAt"),
                cheers = o.optInt("cheers"),
                iCheered = o.optBoolean("iCheered"),
                commentCount = o.optInt("commentCount"),
                lastCommentUser = if (o.isNull("lastCommentUser")) null else o.optString("lastCommentUser"),
                lastCommentText = if (o.isNull("lastCommentText")) null else o.optString("lastCommentText"),
            )
        }
    }

    /** Posts a bean check-in with its photo downscaled to ~800px JPEG. */
    suspend fun postBean(a: Account, bean: Bean): Unit = withContext(Dispatchers.IO) {
        val photoB64 = bean.photoPath?.let { path ->
            BitmapFactory.decodeFile(path)?.let { bmp ->
                val scale = minOf(1f, 800f / maxOf(bmp.width, bmp.height))
                val scaled = if (scale < 1f) {
                    Bitmap.createScaledBitmap(bmp, (bmp.width * scale).toInt(), (bmp.height * scale).toInt(), true)
                } else bmp
                val out = ByteArrayOutputStream()
                // keep PNG for cutouts so transparency survives into the feed
                if (BagCropper.isCutout(path)) scaled.compress(Bitmap.CompressFormat.PNG, 100, out)
                else scaled.compress(Bitmap.CompressFormat.JPEG, 82, out)
                Base64.encodeToString(out.toByteArray(), Base64.NO_WRAP)
            }
        }
        request(a.serverUrl, "/beans", "POST", a.token, JSONObject().apply {
            put("name", bean.name)
            put("roaster", bean.roaster)
            put("origin", bean.origin)
            put("variety", bean.variety)
            put("process", bean.process)
            put("notes", bean.notes)
            put("rating", bean.rating.toDouble())
            photoB64?.let { put("photo_b64", it) }
        })
    }

    /** Fetch + cache a feed photo (keyed by post id) for display. */
    suspend fun fetchPhoto(context: Context, post: FeedPost): File? = withContext(Dispatchers.IO) {
        val url = post.photoUrl ?: return@withContext null
        val dir = File(context.cacheDir, "feed").apply { mkdirs() }
        val ext = if (url.endsWith(".png")) "png" else "jpg"
        val f = File(dir, "${post.id}.$ext")
        if (f.exists() && f.length() > 0) return@withContext f
        runCatching {
            val conn = URL(url).openConnection() as HttpURLConnection
            conn.connectTimeout = 6000
            conn.readTimeout = 15000
            conn.inputStream.use { input -> f.outputStream().use { input.copyTo(it) } }
            conn.disconnect()
            f
        }.getOrNull()
    }
}
